server/controllers/question.js

```diff
@@ -0,0 +1,57 @@
+const Question = require("../models/question");
+const User = require("../models/user");
+var ObjectId = require("mongoose").Types.ObjectId;
+
+module.exports.getSingleQuestion = async (req, res, next) => {
+  const { difficulty } = req.body;
+  const { user } = req.user;
+  const userId = user._id;
+  console.log("user id:", userId);
+
+  try {
+    const user = await User.findById(userId);
+    const question = await Question.findOne({
+      difficulty,
+      _id: { $nin: user.questionIds },
+    });
+
+    if (question) {
+      await User.findOneAndUpdate(
+        { _id: userId },
+        { $push: { questionIds: question._id } }
+      );
+      res.status(200).json(question);
+    } else {
+      let allQuestionIdsWithSameDifficultyLevel = await Question.find(
+        {
+          difficulty,
+        },
+        { title: 0, body: 0, solution: 0, difficulty: 0, _id: 1 }
+      );
+      allQuestionIdsWithSameDifficultyLevel = allQuestionIdsWithSameDifficultyLevel.map(
+        (idObj) => idObj._id
+      );
+
+      await User.findOneAndUpdate(
+        { _id: userId },
+        {
+          $pull: {
+            questionIds: { $in: allQuestionIdsWithSameDifficultyLevel },
+          },
+        }
+      );
+
+      const newQuestion = await Question.findOne({
+        difficulty,
+      });
+      // console.log("new quest:", newQuestion);
+      await User.findOneAndUpdate(
+        { _id: userId },
+        { $push: { questionIds: newQuestion._id } }
+      );
+      res.status(200).json(newQuestion);
```

Comment:
Can you clarify what this block is doing? My understanding is that it's removing all questions of a certain difficulty from the user and then adding one of that difficulty. What is the use case for this?

Reply:
each time a user send a req to get a new question, we should send a question that user has never seen (that is why we save the id of each question into the questionIds field of user in the db). so the above code is for the case a user has seen all the questions from the same difficulty level (that we have in the db) & there is no new question on that difficulty level to be sent to the user for the next req. In that case, we removed the ids of those questions (all questions with that level) from questionIds field of user, so that user can start to fetch question again but this time, inevitably user will get question that has seen before because he already has seen all the questions of that specific level that we have in our db.